app/include/drivers/behavior.h

```diff
@@ -20,7 +20,7 @@
  */
 
 typedef int (*behavior_keymap_binding_callback_t)(struct device *dev, u32_t position, u32_t param1,
-                                                  u32_t param2);
+                                                  u32_t param2, s64_t timestamp);
```

Comment:
It seems like we're starting to reach the point of "too much" in terms of parameters. I wonder if perhaps instead of these proliferation of parameters, if we need a `struct zmk_key_position_event_data` struct that contains all the data? Then the event definition itself would be just the header and then the data? I start this approach for some of the layer state events I have half completed.

Thoughts?

Reply:
As in ... this?
```
struct zmk_key_position_event_data {
    u32_t position;
    s64_t timestamp;
};
```
or?

Reply:
Yes, at a minimum. I do wonder if we should figure out a way to "bundle" the two binding parameters too?

Reply:
Parameters as an inner struct or fixed-size array?  What's the cleanest approach to handling one-param and two-param bindings in this context?

Reply:
I would still expect a fixed two fields for each possible param, that would be `0` and ignored by behaviors that don't use them.

Reply:
yes, I had the same thoughts when developing this. Lets pass by value for now to make memory mgmt simple.

Reply:
```
struct zmk_key_position_event_data {
    32_t param1,
    32_t param2,
    u32_t position;
    s64_t timestamp;
};
```

Reply:
Suggest `position` and `timestamp` at the start of the `struct` as those are the least likely to change in the future.

Reply:
The overlap with two existing structs is remarkable. Although the first contains a char* instead of a device*, and the second is allocated on the stack and has the wrong lifetime...

```
struct zmk_behavior_binding {
    char *behavior_dev;
    u32_t param1;
    u32_t param2;
};
```
```
struct position_state_changed {
    struct zmk_event_header header;
    u32_t position;
    bool state;
    s64_t timestamp;
};
```

Reply:
Yeah, so the other issue here is there's the general "event data", and then there's specific "binding info" that should be passed to the behavior we're activating.

I think perhaps we end up with something like:

```
struct zmk_position_state_changed_data {
  u32_t position;
  bool state;
  s64_t timestamp;
}

struct position_state_changed {
  struct zmk_event_header header;
  struct zmk_position_state_changed_data data;
}
```

And then we can pass two things to the behavior, the position state changed data struct, and the binding struct.

Thoughts?

Reply:
I'm now implementing the following struct. Not happy with the naming.

```
struct behavior_keymap_binding_data {
    int layer;
    u32_t position;
    s64_t timestamp;
    u32_t param1;
    u32_t param2;
};
```

Reply:
@okke-formsma Regarding naming `zmk_behavior_binding_event_data` maybe?

Reply:
There's also a behavior binding event for sensors which acts slightly different, I'll see if we can use the same data.

Reply:
I still am not convinced we don't need two structs, both passed to the behavior callbacks, one for the event data itself, separate from the specific behavior being invoked, and the other for the specific binding data that is associated w/ this specific behavior being invoked.

Thoughts?

Reply:
After considering all ideas, I've decided to implement the following:

```
struct zmk_behavior_binding {
    char *behavior_dev;
    u32_t param1;
    u32_t param2;
};

struct zmk_behavior_binding_data {
    int layer;
    u32_t position;
    s64_t timestamp;
};
```

reasoning:
* re-uses existing `struct zmk_behavior_binding`
* "runtime" information stored in `struct zmk_behavior_binding_data`
* not using the position event because this way we can extend the data to include things like the layer information.